pkg/compactor/block_upload_test.go

```diff
@@ -0,0 +1,669 @@
+			setUpBucketMock: func(bkt *bucket.ClientMock) {
+				setUpPartialBlock(bkt)
+				pth := path.Join(tenantID, blockID, fmt.Sprintf("uploading-%s", block.MetaFilename))
+				bkt.MockUpload(pth, nil)
+			},
+			meta: &metadata.Meta{
+				BlockMeta: tsdb.BlockMeta{
+					ULID:    bULID,
+					MinTime: 0,
+					MaxTime: 1000,
+				},
+				Thanos: metadata.Thanos{
+					Labels: map[string]string{
+						mimir_tsdb.CompactorShardIDExternalLabel: "test",
+					},
+					Files: []metadata.File{
+						{
+							RelPath: block.MetaFilename,
+						},
+						{
+							RelPath:   "index",
+							SizeBytes: 1,
+						},
+						{
+							RelPath:   "chunks/000001",
+							SizeBytes: 1024,
+						},
+					},
+				},
+			},
+		},
+		{
+			name:     "failure checking for complete block",
+			tenantID: tenantID,
+			blockID:  blockID,
+			setUpBucketMock: func(bkt *bucket.ClientMock) {
+				bkt.MockExists(path.Join(tenantID, blockID, block.MetaFilename), false, fmt.Errorf("test"))
+			},
+			expInternalServerError: true,
+		},
+		{
+			name:     "complete block already exists",
+			tenantID: tenantID,
+			blockID:  blockID,
+			setUpBucketMock: func(bkt *bucket.ClientMock) {
+				bkt.MockExists(path.Join(tenantID, blockID, block.MetaFilename), true, nil)
+			},
+			expConflict: "block already exists in object storage",
+		},
+		{
+			name:     "failure uploading meta file",
+			tenantID: tenantID,
+			blockID:  blockID,
+			setUpBucketMock: func(bkt *bucket.ClientMock) {
+				setUpPartialBlock(bkt)
+				pth := path.Join(tenantID, blockID, fmt.Sprintf("uploading-%s", block.MetaFilename))
+				bkt.MockUpload(pth, fmt.Errorf("test"))
+			},
+			meta:                   &validMeta,
+			expInternalServerError: true,
+		},
+		{
+			name:     "valid request",
+			tenantID: tenantID,
+			blockID:  blockID,
+			setUpBucketMock: func(bkt *bucket.ClientMock) {
+				setUpPartialBlock(bkt)
+				pth := path.Join(tenantID, blockID, fmt.Sprintf("uploading-%s", block.MetaFilename))
+				bkt.MockUpload(pth, nil)
+			},
+			meta: &validMeta,
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			var bkt bucket.ClientMock
+			if tc.setUpBucketMock != nil {
+				tc.setUpBucketMock(&bkt)
+			}
+
+			cfgProvider := newMockConfigProvider()
+			if !tc.noRetention {
+				cfgProvider.userRetentionPeriods[tenantID] = time.Second
+			} else {
+				cfgProvider.userRetentionPeriods[tenantID] = 0
+			}
+			c := &MultitenantCompactor{
+				logger:       log.NewNopLogger(),
+				bucketClient: &bkt,
+				cfgProvider:  cfgProvider,
+			}
+			var rdr io.Reader
+			if tc.body != "" {
+				rdr = bytes.NewReader([]byte(tc.body))
+			} else if tc.meta != nil {
+				buf := bytes.NewBuffer(nil)
+				require.NoError(t, json.NewEncoder(buf).Encode(tc.meta))
+				rdr = buf
+			}
+			r := httptest.NewRequest(http.MethodPost, fmt.Sprintf("/api/v1/upload/block/%s", tc.blockID), rdr)
+			if tc.tenantID != "" {
+				r.Header.Set(user.OrgIDHeaderName, tenantID)
+			}
```

Comment:
This made me realize that we extract tenant in a wrong way. We should use `tenant.TenantID` in the handlers, and inject it into the context here. Our authentication middleware is responsible for checking the header.